File: app/src/main/java/com/pr0gramm/app/ui/SingleViewAdapter.java
package com.pr0gramm.app.ui;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;

/**
 */
public class SingleViewAdapter<V extends View> extends SingleValueAdapter<Void, V> {
    private final ViewFactory<V> viewFactory;

    public SingleViewAdapter(ViewFactory<V> viewFactory) {
        super(null);
        this.viewFactory = viewFactory;
    }

    @Override
    protected V createView(Context context, ViewGroup parent) {
        return viewFactory.newView(context);
    }

    @Override
    protected void bindView(V view, Void object) {
    }

    public interface ViewFactory<V> {
        V newView(Context context);
    }

    public static <V extends View> SingleViewAdapter<V> ofView(V view) {
        return new SingleViewAdapter<>(context -> view);
    }

    public static <V extends View> SingleViewAdapter<V> of(ViewFactory<V> factory) {
        return new SingleViewAdapter<>(factory);
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/SingleViewAdapter.kt
package com.pr0gramm.app.ui

import android.content.Context
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes

/**
 */
abstract class SingleViewAdapter : androidx.recyclerview.widget.RecyclerView.Adapter<androidx.recyclerview.widget.RecyclerView.ViewHolder>() {
    init {
        setHasStableIds(true)
    }

    override fun getItemCount(): Int = 1

    override fun getItemId(position: Int): Long = 1

    override fun onBindViewHolder(holder: androidx.recyclerview.widget.RecyclerView.ViewHolder, position: Int) {
        // do nothing
    }

    class ViewHolder(view: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(view)

    companion object {
        fun ofLayout(@LayoutRes layoutId: Int): SingleViewAdapter {
            return of { context -> View.inflate(context, layoutId, null) }
        }

        inline fun of(crossinline factory: (Context) -> View): SingleViewAdapter {
            return object : SingleViewAdapter() {
                override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): androidx.recyclerview.widget.RecyclerView.ViewHolder {
                    return ViewHolder(factory(parent.context))
                }
            }
        }
    }
}
